File: sdks/typescript/src/core/arbitration.ts
/**
 * DCP-07 v2.0 Arbitration & Jurisprudence.
 *
 * Implements arbitration panels (reusing M-of-N governance ceremony pattern),
 * resolution submission, jurisprudence bundle creation, and precedent lookup.
 */

import type { CompositeKeyPair } from './composite-ops.js';
import { compositeSign } from './composite-ops.js';
import type { AlgorithmRegistry } from './crypto-registry.js';
import { canonicalizeV2 } from './canonicalize.js';
import { DCP_CONTEXTS } from './domain-separation.js';
import type {
  AuthorityLevel,
  ArbitrationResolution,
  JurisprudenceBundle,
} from '../types/v2.js';

export interface ArbitrationPanel {
  arbitrator_ids: string[];
  threshold: number;
  created_at: string;
}

/**
 * Create an arbitration panel (M-of-N pattern from governance.ts).
 */
export function createArbitrationPanel(
  arbitratorIds: string[],
  threshold: number,
): ArbitrationPanel {
  if (arbitratorIds.length < threshold) {
    throw new Error(
      `Arbitration panel: need at least ${threshold} arbitrators, got ${arbitratorIds.length}`,
    );
  }
  if (threshold < 1) {
    throw new Error('Arbitration panel: threshold must be >= 1');
  }

  return {
    arbitrator_ids: arbitratorIds,
    threshold,
    created_at: new Date().toISOString(),
  };
}

/**
 * Submit a resolution for a dispute.
 */
export async function submitResolution(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    dispute_id: string;
    session_nonce: string;
    arbitrator_ids: string[];
    resolution: string;
    binding: boolean;
    precedent_references?: string[];
  },
): Promise<ArbitrationResolution> {
  const payload = {
    dcp_version: '2.0' as const,
    dispute_id: params.dispute_id,
    session_nonce: params.session_nonce,
    arbitrator_ids: params.arbitrator_ids,
    resolution: params.resolution,
    binding: params.binding,
    precedent_references: params.precedent_references,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Dispute, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Build a jurisprudence bundle from a resolved dispute.
 */
export async function buildJurisprudenceBundle(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    jurisprudence_id: string;
    session_nonce: string;
    dispute_id: string;
    resolution_id: string;
    category: string;
    precedent_summary: string;
    applicable_contexts: string[];
    authority_level: AuthorityLevel;
  },
): Promise<JurisprudenceBundle> {
  const payload = {
    dcp_version: '2.0' as const,
    jurisprudence_id: params.jurisprudence_id,
    session_nonce: params.session_nonce,
    dispute_id: params.dispute_id,
    resolution_id: params.resolution_id,
    category: params.category,
    precedent_summary: params.precedent_summary,
    applicable_contexts: params.applicable_contexts,
    authority_level: params.authority_level,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Dispute, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Lookup precedent from a jurisprudence collection.
 */
export function lookupPrecedent(
  jurisprudence: JurisprudenceBundle[],
  category: string,
  context?: string,
): JurisprudenceBundle[] {
  return jurisprudence.filter((j) => {
    if (j.category !== category) return false;
    if (context && !j.applicable_contexts.includes(context)) return false;
    return true;
  });
}

File: sdks/python/dcp_ai/v2/arbitration.py
"""
DCP-07 v2.0 Arbitration & Jurisprudence — Python port.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS
from dcp_ai.v2.lifecycle import _utc_now


@dataclass
class ArbitrationPanel:
    arbitrator_ids: list[str]
    threshold: int
    created_at: str


def create_arbitration_panel(arbitrator_ids: list[str], threshold: int) -> ArbitrationPanel:
    """Create an arbitration panel (M-of-N ceremony pattern)."""
    if threshold < 1:
        raise ValueError("Arbitration panel: threshold must be >= 1")
    if len(arbitrator_ids) < threshold:
        raise ValueError(
            f"Arbitration panel: need at least {threshold} arbitrators, "
            f"got {len(arbitrator_ids)}"
        )
    return ArbitrationPanel(
        arbitrator_ids=arbitrator_ids,
        threshold=threshold,
        created_at=_utc_now(),
    )


async def submit_resolution(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    dispute_id: str,
    session_nonce: str,
    arbitrator_ids: list[str],
    resolution: str,
    binding: bool,
    precedent_references: list[str] | None = None,
) -> dict[str, Any]:
    """Submit a resolution for a dispute."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "dispute_id": dispute_id,
        "session_nonce": session_nonce,
        "arbitrator_ids": arbitrator_ids,
        "resolution": resolution,
        "binding": binding,
        "precedent_references": precedent_references,
        "timestamp": _utc_now(),
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Dispute"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


async def build_jurisprudence_bundle(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    jurisprudence_id: str,
    session_nonce: str,
    dispute_id: str,
    resolution_id: str,
    category: str,
    precedent_summary: str,
    applicable_contexts: list[str],
    authority_level: str,
) -> dict[str, Any]:
    """Build a jurisprudence bundle from a resolved dispute."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "jurisprudence_id": jurisprudence_id,
        "session_nonce": session_nonce,
        "dispute_id": dispute_id,
        "resolution_id": resolution_id,
        "category": category,
        "precedent_summary": precedent_summary,
        "applicable_contexts": applicable_contexts,
        "authority_level": authority_level,
        "timestamp": _utc_now(),
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Dispute"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


def lookup_precedent(
    jurisprudence: list[dict[str, Any]],
    category: str,
    context: str | None = None,
) -> list[dict[str, Any]]:
    """Filter a jurisprudence collection by category (and optional context)."""
    out: list[dict[str, Any]] = []
    for entry in jurisprudence:
        if entry.get("category") != category:
            continue
        if context is not None and context not in entry.get("applicable_contexts", []):
            continue
        out.append(entry)
    return out
